integration/cluster.go

```diff
@@ -1272,3 +1274,116 @@ type grpcAPI struct {
 	// Election is the election API for the client's connection.
 	Election epb.ElectionClient
 }
+
+// GetLearnerMembers returns the list of learner members in cluster using MemberList API.
+func (c *ClusterV3) GetLearnerMembers() ([]*pb.Member, error) {
+	cli := c.Client(0)
+	resp, err := cli.MemberList(context.Background())
+	if err != nil {
+		return nil, fmt.Errorf("failed to list member %v", err)
+	}
+	var learners []*pb.Member
+	for _, m := range resp.Members {
+		if m.IsLearner {
+			learners = append(learners, m)
+		}
+	}
+	return learners, nil
+}
+
+// AddAndLaunchLearnerMember creates a leaner member, adds it to cluster
+// via v3 MemberAdd API, and then launches the new member.
+func (c *ClusterV3) AddAndLaunchLearnerMember(t testing.TB) {
+	m := c.mustNewMember(t)
+	m.isLearner = true
+
+	scheme := schemeFromTLSInfo(c.cfg.PeerTLS)
+	peerURLs := []string{scheme + "://" + m.PeerListeners[0].Addr().String()}
+
+	cli := c.Client(0)
+	_, err := cli.MemberAdd(context.Background(), peerURLs, m.isLearner)
+	if err != nil {
+		t.Fatalf("failed to add learner member %v", err)
+	}
+
+	m.InitialPeerURLsMap = types.URLsMap{}
+	for _, mm := range c.Members {
+		m.InitialPeerURLsMap[mm.Name] = mm.PeerURLs
+	}
+	m.InitialPeerURLsMap[m.Name] = m.PeerURLs
+	m.NewCluster = false
+
+	if err := m.Launch(); err != nil {
+		t.Fatal(err)
+	}
+
+	c.Members = append(c.Members, m)
+
+	c.waitMembersMatch(t)
+}
+
+// getMembers returns a list of members in cluster, in format of etcdserverpb.Member
+func (c *ClusterV3) getMembers() []*pb.Member {
+	var mems []*pb.Member
+	for _, m := range c.Members {
+		mem := &pb.Member{
+			Name:       m.Name,
+			PeerURLs:   m.PeerURLs.StringSlice(),
+			ClientURLs: m.ClientURLs.StringSlice(),
+			IsLearner:  m.isLearner,
+		}
+		mems = append(mems, mem)
+	}
+	return mems
+}
+
+// waitMembersMatch waits until v3rpc MemberList returns the 'same' member info as
+// the local c.Members, which is used by integration testing. With the exception that
+// c.Members does not have info on Member.ID, which is generated in testing cluster.
+//
+// Note:
+// A successful match means the clientURLs are also matched. This means server has already
+// finished publishing its s.attributes to cluster. Publish attributes is a cluster-wide
+// write request. Therefore any raft entries prior to this have already been applied.
+// To a newly added member (to an existing cluster), this means that it has already
+// finished applying config changes of adding existing members and itself (replicated
+// from leader) to local raft node and server's membership.RaftCluster.
+//
+// Special note on learner member:
+// Learner member is added via v3rpc MemberAdd API (as of v3.4). When starting the learner
+// member, initial membership.RaftCluster created by peerURLs map does not have info on
```

Comment:
I see the same thing during the test. and sometimes new learner node become followers accidentally when startup.

Reply:
Shouldn't learner become follower?

Reply:
Can you provide more detail? I feel we might need to make some changes to learner node startup process. Currently there is a small time period during startup, where learner node's `Member.RaftAttributes.IsLearner` is `false`.

Reply:
ok

Reply:
I was thinking about this issue. The problem is, any changes to `Member.RaftAttributes` has to go through raft consensus, as it is by definition a configuration change. So when starting the learner node, while it is starting and catching up on applying the raft log entries (especially the `ConfChangeAddLearnerNode` entry that was used to add this learner to the cluster), there will be no update to learner's local `Member.RaftCluster`, specifically the learner's `Member.RaftAttributes.IsLearner` field.

Maybe during the very beginning of learner's startup, it is fine to mark `IsLearner` field to true without going through Raft consensus. When learner joins an existing cluster, it will get existing cluster information from its peer, so it knows itself is learner.

Reply:
> Shouldn't learner become follower?

 I think if it become follower, it can vote... maybe change to leader

Reply:
> > Shouldn't learner become follower?
> 
> I think if it become follower, it can vote... maybe change to leader

FYI:
Learner will become a follower when starting.
https://github.com/jingyih/etcd/blob/a621d807f061e1dd635033a8d6bc261461429e27/raft/node.go#L200-L204

And, it will not vote as follower.
https://github.com/jingyih/etcd/blob/a621d807f061e1dd635033a8d6bc261461429e27/raft/raft.go#L932-L938

And, it cannot start a campaign to become leader.
https://github.com/jingyih/etcd/blob/a621d807f061e1dd635033a8d6bc261461429e27/raft/raft.go#L670-L676
https://github.com/jingyih/etcd/blob/a621d807f061e1dd635033a8d6bc261461429e27/raft/raft.go#L1393-L1396

Reply:
We persist raft attributes in v3 store:

```go
func mustSaveMemberToBackend(be backend.Backend, m *Member) {
	mkey := backendMemberKey(m.ID)
	mvalue, err := json.Marshal(m)
        ...
```

@jingyih @WIZARD-CXY Can you help understand why restart won't reload learner state? Don't we persist learner member attributes?

Reply:
@gyuho
This is not restarting a node, but rather starting a new learner node.

When adding learner node to an exsiting cluster, we do:

1) `etcdctl member add --learner`, so that all existing members:
1.1) append `ConfChangeAddLearnerNode` entry to their raft log.
1.2) apply the config change to their local raft node.
1.3) apply the config change to server cluster `Etcdserver.cluster`. This is where the member info are saved to v3 store.

2) Start etcd binary for the new learner node. As a new node, it does not have any member information stored in v3 store yet. After been started, it will take some time for the new learner's raft node to catch up on applying the raft log entries, including the `ConfChangeAddLearnerNode` mentioned above. Until then, v3 store and `Etcdserver.cluster` on the newly added learner node will not have learner information. i.e., the learner does not know itself is learner.

Does this make sense to you?